### PR TITLE
Issue 1053: Update django-admin-508 for a11y fixes

### PR DESCRIPTION
**Pull request by jorgegonzalez**
### Summary of Changes

This PR updates `django-adming-508` so as to fix several outstanding a11y issues in Django Admin.

List of fixes: <https://github.com/raft-tech/django-admin-508/blob/main/CHANGELOG.md#003-2021-07-21>

_Addresses issue_ #1053

**Completed Tasks:**
- [x] Evaluate a11y fixes coming from fork 
- - [x] Add outstanding contrast fixes if needed (See Task 4 on hack.md)
- [x] Add dynamic labels to checkbox elements (See Task 1 on hack.md)
- [x] Ensure errors and other state changes are read by screenreaders (See Task 3 on hack.md)
- [x] Make heading styles and associated markup consistent across Django Admin (See Task 6 on hack.md)
- [x] Retain the tooltips, implement accessible tooltip ARIA markup (See Task 2 on hack.md)

**Supporting Documentation:**
[https://hackmd.io/@mreiter/ryuHXhbuu](https://hackmd.io/@mreiter/ryuHXhbuu
) Contains full task list for a11y fixes in support of #892 

### How to Test
_List the steps to test the PR_

1. Log into Django Admin
2. Confirm the fixes noted in the tasks above

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
	- [x] Django admin is fully visually accessible (e.g. color, contrast) to at least a WCAG 2.0 AA standard.
	- [x] Django admin is more accessible to screen reader only users (markup conventions, all alerts / state / value updates are read) 
	- [x] Testing Checklist has been run and all tests pass

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

Does not apply; this PR introduces only a dependency update, and no changes to backend or frontend application code.

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

Does not apply, no code changes included.

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

Does not apply, existing dependency is being updated to a newer version.

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
	- No issues found

### CODE REVIEW

File: tdrs-backend/Pipfile [hunk 1]
```diff
@@ -44,7 +44,7 @@ requests = "~=2.25.0"
 django-admin-logs = "~=1.0.1"
 drf-yasg = "~=1.20.0"
 wait-for-it = "==2.2.0"
-django-admin-508 = "==0.0.2"
+django-admin-508 = "==0.0.3"
```
Comment (jorgegonzalez) [inline tdrs-backend/Pipfile:47]: Merely updating our dependency will get us the fixes applied automatically.

**Comment by jorgegonzalez** [comment]
I added a [changelog](https://github.com/raft-tech/django-admin-508/blob/main/CHANGELOG.md#003-2021-07-21) with a detailed list of fixes with commit links in this version.

**Review by jtwillis92 — APPROVED** [review]
LGTM! I went through the referenced changelog and confirmed that the new JS files are loaded and the appropriate accessibility tweaks are in place on the deployed a11y site.

As a nit, I wonder if the aria label added [here](https://github.com/raft-tech/django-admin-508/commit/b14237d8b64e54ce2b0d5590d929c15eaf5ce99a#diff-5e9ea7bce3b74d9d6e08ea5474d3821e2ffc7248e3f20d997e6c0e19a7b0d667R13) should use the `username` field instead of the ID. The resulting aria-label isn't super human friendly:
```
aria-label="Row ece149ce-4eec-45a7-b7c6-1e0366f91c76"
```

**Comment by jtwillis92** [comment]
> As a nit, I wonder if the aria label added [here](https://github.com/raft-tech/django-admin-508/commit/b14237d8b64e54ce2b0d5590d929c15eaf5ce99a#diff-5e9ea7bce3b74d9d6e08ea5474d3821e2ffc7248e3f20d997e6c0e19a7b0d667R13) should use the `username` field instead of the ID. The resulting aria-label isn't super human friendly:
> 
> ```
> aria-label="Row ece149ce-4eec-45a7-b7c6-1e0366f91c76"
> ```

Well, I guess we can't always use the `username` since not all models will have that field. I'm not sure if there is an alternative there.

**Comment by jorgegonzalez** [comment]
Hey good catch, what table were you looking at with this? We could do something like `checkbox?.name ?? checkbox?.value` here actually

**Comment by jtwillis92** [comment]
> Hey good catch, what table were you looking at with this? We could do something like `checkbox?.name ?? checkbox?.value` here actually

It was the users table, on the list page where you can select multiple.

**Comment by jorgegonzalez** [comment]
@jtwillis92 I added a new version with that fix, thanks!

**Review by reitermb — APPROVED** [review]
Raft a11y review: 

> Add outstanding contrast fixes if needed (See Task 4 on hack.md)

Completed with a contrast fix for pill shaped Add/History buttons on various pages. I'll note that some automated checkers will still flag certain elements for review of color, but that's down to how they parse rather than what the actual colors are.

> Add dynamic labels to checkbox elements

All checkbox elements should now have a unique label that gets read by screenreaders. Note that not all of these dynamic labels are yet pointing at human-readable attributes. e..g. Alabama might be read as an ID like "1429" rather than Alabama. This will be addressed in #1054. 

> Ensure errors and other state changes are read by screenreaders

Form errors do now get read via aria alerts or labeling markup when they exist/occur. I did flag some instances of focus being grabbed by the wrong element on the page (interrupting the error) in Chrome on MacOS—still investigating this / evaluating potential focus-related fixes for #1054 or a separate follow-on. We may also make the copy of certain errors clearer in a later follow-on ticket, e.g. things like "First Name cannot be blank" instead of "Field cannot be blank".

> Retain the tooltips, implement accessible tooltip ARIA markup

Tooltips are now scannable and screen reader readable in Safari/VoiceOver. In Chrome we discovered that there's a general lack of support for the title attribute in tooltip interactions. We're investigating whether there's a focus improvement we could loop into #1054 to add greater cross-browser support to the tooltips and make them easier to discover.

**Review by abottoms-coder — APPROVED** [review]

**Comment by jorgegonzalez** [comment]
> Completed with a contrast fix for pill shaped Add/History buttons on various pages

Screen Shot 2021-08-13 at 4 01 45 PM
Screen Shot 2021-08-13 at 4 01 08 PM

> All checkbox elements should now have a unique label that gets read by screenreaders. Note that not all of these dynamic labels are yet pointing at human-readable attributes

Screen Shot 2021-08-13 at 4 09 15 PM

Screen Shot 2021-08-13 at 4 09 35 PM

> Tooltips are now scannable and screen reader readable in Safari/VoiceOver

Screen Shot 2021-08-13 at 4 21 54 PM

Screen Shot 2021-08-13 at 4 21 44 PM


> Form errors do now get read via aria alerts or labeling markup when they exist/occur

Screen Shot 2021-08-13 at 4 24 23 PM

**Review by ADPennington — APPROVED** [review]
thanks @jorgegonzalez : @abottoms-coder  ready to merge. 

Notes for follow-up below:

- this PR includes work to address tasks 1, 2, 3, 4 and 6, but there are a few things to check for and wrap up for tasks 1,2, and 3 as noted [here](https://github.com/raft-tech/TANF-app/pull/1137#pullrequestreview-727691024). @reitermb  also will add to the hackmd to document progress on completion of each of the tasks for our future reference. 
- A couple of enhancements @ttran-hub and I think would be worth considering down the road (V3+): (1) the pill-shaped buttons are a grey when hovered over but, for consistency with other hover actions and to support better contrast, worth changing this color. (2) the hover over color for many of the links are not very distinctive from the primary color (hardly noticeable). Miles will be creating a ticket that includes a broader list of potential a11y enhancements to be considered further down the road. 

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] Evaluate a11y fixes coming from fork
  + [x]  Add outstanding contrast fixes if needed (See Task 4 on hack.md)
  + [x] Add dynamic labels to checkbox elements (See Task 1 on hack.md)
  + [x] Ensure errors and other state changes are read by screenreaders (See Task 3 on hack.md)
  + [x] Make heading styles and associated markup consistent across Django Admin (See Task 6 on hack.md)
  + [x] Retain the tooltips, implement accessible tooltip ARIA markup (See Task 2 on hack.md)

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
_manually checked code changes on repo for dependency. checked screenshots and changes on deployed site_
+ Are code coverage minimums met?
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
+ [ ] Does this PR complete the epic?
_no, but #1054 will_
+ [x] Are links included to any other gov-approved PRs associated with epic?

_PRs associated with #1054 and #973_
+ [x] Does PR include documentation for Raft's a11y review?
https://github.com/raft-tech/TANF-app/pull/1137#pullrequestreview-727691024


### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no new issues detected_

**Pull request merged**